azure-iot-device/azure/iot/device/common/pipeline/pipeline_ops_base.py

```diff
@@ -44,10 +51,79 @@ def __init__(self, callback):
             )
         self.name = self.__class__.__name__
-        self.callback = callback
+        self.callbacks = []
         self.needs_connection = False
         self.completed = False
 
+        self.callbacks.append(callback)
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def add_callback(self, callback):
+        self.callbacks.append(callback)
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def complete(self, error=None):
+        if error:
+            logger.error("{}: completing with error {}".format(self.name, error))
+        else:
+            logger.debug("{}: completing without error".format(self.name))
+
+        if self.completed:
+            logger.error("{}: has already been completed!".format(self.name))
+            e = pipeline_exceptions.OperationError(
+                "Attempting to complete an already-completed operation: {}".format(self.name)
+            )
+            # This could happen in a foreground or background thread, so err on the side of caution
+            # and send it to the background handler.
+            handle_exceptions.handle_background_exception(e)
+        else:
+            self.completed = True
+
+            while self.callbacks:
+                if not self.completed:
+                    logger.debug(
+                        "{}: Operation no longer complete! Suspending completion".format(self.name)
+                    )
+                    break
+                callback = self.callbacks.pop()
+                try:
+                    callback(op=self, error=error)
+                except Exception as e:
+                    logger.error(
+                        "Unhandled error while triggering callback for {}".format(self.name)
+                    )
+                    logger.error(traceback.format_exc())
+                    # This could happen in a foreground or background thread, so err on the side of caution
+                    # and send it to the background handler.
+                    handle_exceptions.handle_background_exception(e)
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def uncomplete(self):
+        logger.debug("{}: Uncompleting".format(self.name))
+        self.completed = False
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def spawn_worker_op(self, worker_op_type, **kwargs):
+        logger.debug("{}: creating worker op of type {}".format(self.name, worker_op_type.__name__))
+
+        # CT-TODO: does this need to use a weakref?
+
+        @pipeline_thread.runs_on_pipeline_thread
+        def on_worker_op_complete(op, error):
+            logger.debug("{}: Worker op ({}) has been completed".format(self.name, op.name))
+            self.complete(error=error)
+
+        if "callback" in kwargs:
+            provided_callback = kwargs["callback"]
+            kwargs["callback"] = on_worker_op_complete
+            worker_op = worker_op_type(**kwargs)
+            worker_op.add_callback(provided_callback)
+        else:
```

Comment:
I don't think I like this code.  You're making some big assumptions here about the nature of the pre-existing callback, and you're making for a confusing flow if there is a pre-existing callback.  If I understand the code, I think completing a worker with a pre-existing callback will first call all the callbacks on the parent op, then, after all that is done, it will call the pre-existing callback.  This doesn't seem useful.

Reply:
I believe you are mistaken, unless I am mistaken.

What this does is, if the user provides a callback as kwarg for creating a worker op, that callback will be added to the callback_stack AFTER the callback that completes the original op. And because it's a stack, that means that this user provided callback will be called FIRST and BEFORE completing the parent op that spawned the worker. Completing the parent op is ALWAYS last.

---
In reply to: [345996131](https://github.com/Azure/azure-iot-sdk-python/pull/392#discussion_r345996131) [](ancestors = 345996131)

Reply:
OK, I can see the utility in this.  As long as there's a UT to validate the ordering, I'm OK with this.  

---
In reply to: [346547761](https://github.com/Azure/azure-iot-sdk-python/pull/392#discussion_r346547761) [](ancestors = 346547761,345996131)

Reply:
Yes, I've double checked, there are tests for this, and as a result requirements as well.

---
In reply to: [347011108](https://github.com/Azure/azure-iot-sdk-python/pull/392#discussion_r347011108) [](ancestors = 347011108,346547761,345996131)